lib/AnomalyTree.js

```diff
@@ -90,15 +102,26 @@ AnomalyTree.prototype.depth = function (inputData, path, depth) {
     depth = 0;
   }
   if (depth === 0) {
-    if (!this.predicates.evaluate(inputData, this.fields)) {
+    const preds = this.predicates || predicates(this.tree);
+    if (!preds.evaluate(inputData, this.fields)) {
       return {depth: depth, path: path};
     }
     depth += 1;
   }
-  if (this.children) {
-    len = this.children.length;
+
+  let children = this.tree ? this.tree.children : this.children;
+  if (children) {
+    len = children.length;
     for (index = 0; index < len; index++) {
-      child = this.children[index];
+      if (this.children) { // if this.children, we are non-lazy
```

Comment:
Why don't we use the this.lazy attribute that you just defined? you would not need the comment...

Reply:
you can pass opts.lazy -- when opts.shallow is false, opts.lazy is forced to be false as well because if you are going to create the whole tree hierarchy, you better create the predicates as well at the same time (it is of no use to have the tree hierarchy with no predicates) -- othersiwe, this.lazy gets the value you pass for opts.lazy.